torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1099,6 +1117,286 @@ def call_function(
         return super().call_function(tx, args, kwargs)
 
 
+class CannotConvertRangeToHigherOrder(TorchDynamoException):
+    pass
+
+
+class RangeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    The general idea is that for a simple for loop like
+
+    ```
+    for i in range(10):
+        x = x + 1
+    ```
+
+    We can convert the loop body to a function and the loop itself
+    to a series of calls:
+
+    ```
+    def loop_body(local_0, local_1, local_...):
+        local_1 = local_1 + 1
+        return (local_0, local_1, local_...)
+
+    (i, x) = loop_body(0, x)
+    (i, x) = loop_body(1, x)
+    (i, x) = loop_body(2, x)
+    # ... 7 more times
+    ```
+
+    There are a few benefits to doing this, but mainly we bypass a ton
+    of stuff associated with tracing for loops. That itself provides some speedup.
+
+    Once we implement function compilation caching, or a higher order operator,
+    like jax.lax.fori_loop, for loops will compile in constant time, and perhaps even
+    better than what we currently have.
+    """
+
+    class UnitializedVariable(ConstantVariable):
+        """Special sentinel to indicate values are not yet set. This differs
+        from misc.UnknownVariable in that it's actually known outside of a for loop
+        higher order op.
+        """
+
+        pass
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
+        "RETURN_VALUE",
+        "YIELD_VALUE",
+        "YIELD_FROM",
+        "GET_YIELD_FROM_ITER",
+        "SEND",
+        "MAKE_FUNCTION",
+        # This could be supported in the future.
+        "LOAD_DEREF",
```

Comment:
What about STORE_DEREF/DELETE_DEREF/etc.

Reply:
I think those will be handled by `speculate_subgraph`, but just to be sure I will add them. In theory a lot of things can be handled by `speculate_subgraph`. For example, it makes code in a `try-except` safe for this transformation too. But just to be safe, I will block anything that touches exceptions or closures.